File: routes/lego.py
from __future__ import annotations
import tempfile
from contextlib import suppress
from pathlib import Path

from fastapi import APIRouter, Request, UploadFile, File, Form
from fastapi.responses import JSONResponse

import config
from core.comfyui import ComfyUIClient
from nyx_step import get_user_email
from routes._helpers import submit_and_register

router = APIRouter()
_client = ComfyUIClient()


@router.post("/lego")
async def lego_generate(
    request: Request,
    audio: UploadFile = File(...),
    tags: str = Form(""),
    lyrics: str = Form(""),
    song_name: str = Form("Lego"),
    denoise: float = Form(0.5),
    steps: int = Form(20),
    cfg: float = Form(2.0),
    duration: float = Form(30.0),
    seed: int = Form(0),
    bpm: int = Form(120),
    key: str = Form("C"),
    scale: str = Form("Major"),
    audio_format: str = Form("mp3"),
    audio_quality: str = Form("V0"),
    dit_model: str = Form("sft"),
    sampler_name: str = Form("er_sde"),
    scheduler: str = Form("linear_quadratic"),
):
    user_email = get_user_email(request)
    content = await audio.read()
    if len(content) > config.MAX_UPLOAD_BYTES:
        return JSONResponse({"error": f"File too large (max {config.MAX_UPLOAD_BYTES // 1024 // 1024} MB)"}, status_code=413)

    suffix = Path(audio.filename).suffix
    with tempfile.NamedTemporaryFile(suffix=suffix, delete=False) as f:
        f.write(content)
        tmp = Path(f.name)
    try:
        input_name = _client.copy_to_input(tmp)
    finally:
        with suppress(FileNotFoundError):
            tmp.unlink()

    state = {
        "bpm": bpm, "key": key, "scale": scale,
        "steps": steps, "cfg_scale": cfg, "duration": duration,
        "seed": seed, "lock_seed": seed != 0,
        "audio_format": audio_format, "audio_quality": audio_quality,
        "dit_model": dit_model, "sampler_name": sampler_name, "scheduler": scheduler,
        "generate_audio_codes": True,
    }
    result = _client.build_lego_workflow(input_name, tags, lyrics, state, denoise)
    if "error" in result:
        return JSONResponse({"error": result["error"]}, status_code=400)

    return submit_and_register(
        _client, user_email, result["workflow"], song_name,
        seed=result.get("seed", 0), caption=tags,
        upstream_error_status=502,
        params={"bpm": bpm, "key": key, "scale": scale, "steps": steps,
                "cfg_scale": cfg, "duration": duration},
    )


@router.post("/complete")
async def complete_generate(
    request: Request,
    audio: UploadFile = File(...),
    tags: str = Form(""),
    lyrics: str = Form(""),
    song_name: str = Form("Complete"),
    denoise: float = Form(0.8),
    steps: int = Form(20),
    cfg: float = Form(2.0),
    duration: float = Form(30.0),
    seed: int = Form(0),
    bpm: int = Form(120),
    key: str = Form("C"),
    scale: str = Form("Major"),
    audio_format: str = Form("mp3"),
    audio_quality: str = Form("V0"),
    dit_model: str = Form("sft"),
    sampler_name: str = Form("er_sde"),
    scheduler: str = Form("linear_quadratic"),
):
    user_email = get_user_email(request)
    content = await audio.read()
    if len(content) > config.MAX_UPLOAD_BYTES:
        return JSONResponse({"error": f"File too large (max {config.MAX_UPLOAD_BYTES // 1024 // 1024} MB)"}, status_code=413)

    suffix = Path(audio.filename).suffix
    with tempfile.NamedTemporaryFile(suffix=suffix, delete=False) as f:
        f.write(content)
        tmp = Path(f.name)
    try:
        input_name = _client.copy_to_input(tmp)
    finally:
        with suppress(FileNotFoundError):
            tmp.unlink()

    state = {
        "bpm": bpm, "key": key, "scale": scale,
        "steps": steps, "cfg_scale": cfg, "duration": duration,
        "seed": seed, "lock_seed": seed != 0,
        "audio_format": audio_format, "audio_quality": audio_quality,
        "dit_model": dit_model, "sampler_name": sampler_name, "scheduler": scheduler,
        "generate_audio_codes": True,
    }
    result = _client.build_lego_workflow(input_name, tags, lyrics, state, denoise)
    if "error" in result:
        return JSONResponse({"error": result["error"]}, status_code=400)

    return submit_and_register(
        _client, user_email, result["workflow"], song_name,
        seed=result.get("seed", 0), caption=tags,
        upstream_error_status=502,
        params={"bpm": bpm, "key": key, "scale": scale, "steps": steps,
                "cfg_scale": cfg, "duration": duration},
    )

File: static/lego.js
// ── Lego Mode & Complete Mode ──────────────────────────────────────────────────
function _legoSubmit(mode) {
  const prefix  = mode === "lego" ? "lego" : "complete";
  const endpoint = mode === "lego" ? "/lego" : "/complete";
  const btnLabel = mode === "lego" ? "🧱 Add Element" : "🎼 Complete Track";

  const btn    = document.getElementById(`btn-${prefix}-submit`);
  const status = document.getElementById(`${prefix}-status`);
  const result = document.getElementById(`${prefix}-result`);
  const fileEl = document.getElementById(`${prefix}-file`);

  if (!fileEl.files.length) { status.textContent = "Select an audio file first."; return; }

  btn.disabled = true;
  btn.textContent = "Sending…";
  status.textContent = "Uploading audio…";
  status.style.color = "var(--muted)";
  result.innerHTML = "";

  const s = mwState;
  const form = new FormData();
  form.append("audio",        fileEl.files[0]);
  form.append("tags",         document.getElementById(`${prefix}-tags`).value.trim());
  form.append("lyrics",       prefix === "complete" ? document.getElementById("complete-lyrics").value : "");
  form.append("song_name",    document.getElementById(`${prefix}-song-name`).value || prefix);
  form.append("denoise",      document.getElementById(`${prefix}-denoise`).value);
  form.append("steps",        document.getElementById(`${prefix}-steps`).value);
  form.append("cfg",          document.getElementById(`${prefix}-cfg`).value);
  form.append("duration",     document.getElementById(`${prefix}-duration`).value);
  form.append("bpm",          s.bpm || 120);
  form.append("key",          s.key || "C");
  form.append("scale",        s.scale || "Major");
  form.append("audio_format", s.audio_format || "mp3");
  form.append("audio_quality",s.audio_quality || "V0");
  form.append("dit_model",    s.dit_model || "sft");
  form.append("sampler_name", s.sampler_name || "er_sde");
  form.append("scheduler",    s.scheduler || "linear_quadratic");
  form.append("seed",         (s.lock_seed && s.seed) ? s.seed : 0);

  fetch(endpoint, { method: "POST", body: form })
    .then(r => r.json().then(data => ({ ok: r.ok, data })))
    .then(({ ok, data }) => {
      if (!ok) {
        status.textContent = "Error: " + (data.error || "unknown");
        status.style.color = "var(--error)";
        return;
      }
      status.textContent = `Queued — prompt ${data.prompt_id.slice(0, 8)}…`;
      status.style.color = "var(--accent2)";
    })
    .catch(e => {
      status.textContent = "Error: " + e.message;
      status.style.color = "var(--error)";
    })
    .finally(() => {
      btn.disabled = false;
      btn.textContent = btnLabel;
    });
}

document.getElementById("btn-lego-submit").addEventListener("click", () => _legoSubmit("lego"));
document.getElementById("btn-complete-submit").addEventListener("click", () => _legoSubmit("complete"));
